relay-server/src/utils/envelope_context.rs

```diff
@@ -91,23 +98,40 @@ impl EnvelopeContext {
     /// As opposed to [`new`](Self::new), this does not require a queue permit. This makes it
     /// suitable for unit testing internals of the processing pipeline.
     #[cfg(test)]
-    pub fn standalone(envelope: &Envelope) -> Self {
+    pub fn standalone(envelope: Box<Envelope>) -> Self {
         Self::new_internal(envelope, None)
     }
 
     /// Computes an envelope context from the given envelope and binds it to the processing queue.
     ///
     /// To provide additional scoping, use [`EnvelopeContext::scope`].
-    pub fn new(envelope: &Envelope, slot: SemaphorePermit) -> Self {
+    pub fn new(envelope: Box<Envelope>, slot: SemaphorePermit) -> Self {
         Self::new_internal(envelope, Some(slot))
     }
 
+    /// Returns a reference to the contained [`Envelope`].
+    pub fn envelope(&self) -> &Envelope {
+        self.envelope.as_ref()
+    }
+
+    /// Returns a mutable reference to the contained [`Envelope`].
+    pub fn envelope_mut(&mut self) -> &mut Envelope {
+        self.envelope.as_mut()
+    }
+
+    /// Take the envelope out of the context and replace it with a dummy.
+    ///
+    /// Note that after taking out the envelope, the envelope summary is incorrect.
+    pub(crate) fn take_envelope(&mut self) -> Box<Envelope> {
+        Box::new(self.envelope.take_items())
```

Comment:
Out of curiosity, how does the rust handle the memory allocated to `envelope` here? Since the object is not destroyed, I assume the same amount of memory is saved?

Reply:
The `Box::new` will cause a new allocation. `take_items` will copy all the headers, but the envelope items themselves remain where they are in memory. Only the pointer to those items is moved.